autometa/binning/summary.py

```diff
@@ -0,0 +1,364 @@
+                "size (bp)": dff.length.sum(),
+                "size_pct": dff.length.sum() / bin_df.length.sum() * 100,
+                "N90": fragmentation_metric(dff, quality_measure=0.9),
+                "N50": fragmentation_metric(dff, quality_measure=0.5),
+                "N10": fragmentation_metric(dff, quality_measure=0.1),
+                "length_weighted_gc": length_weighted_gc,
+                "min_GC": dff.GC.min(),
+                "max_GC": dff.GC.max(),
+                "length_weighted_coverage": length_weighted_coverage,
+                "min_coverage": dff.coverage.min(),
+                "max_coverage": dff.coverage.max(),
+                "completeness": completeness,
+                "purity": purity,
+                "num_total_markers": total_markers,
+                f"num_unique_markers (expected {num_expected_markers})": num_markers_present,
+                "num_single_copy_markers": num_single_copy_markers,
+            }
+        )
+    stats_df = pd.DataFrame(stats)
+    stats_df.set_index("cluster", inplace=True)
+    return stats_df
+
+
+def get_metabin_taxonomies(bin_df, ncbi):
+    """Retrieve taxonomies of all clusters recovered from Autometa binning.
+
+    Parameters
+    ----------
+    bin_df : pd.DataFrame
+        Autometa binning table. index=contig, cols=['cluster','length','taxid', *canonical_ranks]
+    ncbi : autometa.taxonomy.ncbi.NCBI instance
+        Autometa NCBI class instance
+
+    Returns
+    -------
+    pd.DataFrame
+        Dataframe consisting of cluster taxonomy with taxid and canonical rank.
+        Indexed by cluster
+    """
+    canonical_ranks = ncbi.CANONICAL_RANKS
+    if "root" in canonical_ranks:
+        canonical_ranks.remove("root")
+    is_clustered = bin_df.cluster.notnull()
+    bin_df = bin_df[is_clustered]
+    tmpfpath = tempfile.mktemp()
+    outcols = ["cluster", "length", "taxid", *canonical_ranks]
+    bin_df[outcols].to_csv(tmpfpath, sep="\t", index=False, header=False)
+    taxonomies = {}
+    with open(tmpfpath) as fh:
+        for line in fh:
+            llist = line.strip().split("\t")
+            cluster = llist[0]
+            length = int(llist[1])
+            taxid = int(llist[2])
+            ranks = llist[3:]
+            for rank, canonical_rank in zip(ranks, canonical_ranks):
+                if rank != "unclassified":
+                    break
+            if cluster not in taxonomies:
+                taxonomies.update({cluster: {canonical_rank: {taxid: length}}})
+            elif canonical_rank not in taxonomies[cluster]:
+                taxonomies[cluster].update({canonical_rank: {taxid: length}})
+            elif taxid not in taxonomies[cluster][canonical_rank]:
+                taxonomies[cluster][canonical_rank].update({taxid: length})
+            else:
+                taxonomies[cluster][canonical_rank][taxid] += length
+    os.remove(tmpfpath)
+    cluster_taxonomies = rank_taxids(taxonomies, ncbi)
```

Comment:
Have you double checked to see if this gives the same results as the old script?

Reply:
I wrote the tempfile as an easy way to follow the previous construction of the same dictionary that is supplied in the old script. The `rank_taxids` function expects the same data structure. There is probably a way to generate this data structure from the pandas dataframe. But in the interest of time, I opted for writing to a tempfile and constructing the dict while parsing. 

Basically if we would like to remove the `tempfile` requirement, we need to go from:

```python
old = pd.DataFrame(index=range(0, n_contigs-1), cols=["cluster", "length", "taxid", *canonical_ranks])

new = {cluster: {kingdom: {2: 123123, ...}, {phylum: {1224}, ...}, ...}
```